### PR TITLE
Port DeepSeek Sparse Attention to `MambaModel`

### PR DESCRIPTION
**Pull request by janEbert**
# What does this PR do ?

Make experimental DeepSeek Sparse Attention (DSA) available to `MambaModel`. For now, we error out when a user tries to use both standard Attention and DSA. It's not supported right now to have different Attention flavors in the same hybrid model, this requires broader restructuring.

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [X] I have added relevant unit tests
- [X] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/ssm/mamba_hybrid_layer_allocation.py [hunk 1]
```diff
@@ -28,10 +28,11 @@ class Symbols:
 
     MAMBA = "M"
     ATTENTION = "*"
+    DSA_ATTENTION = "S"
```
Comment (duncanriach) [inline megatron/core/ssm/mamba_hybrid_layer_allocation.py]: Wondering if 'S' should be reserved for sliding-window attention. Wondering if this should be 'D'. Of course, these choices are arbitrary and hopefully ultimately temporary.

File: megatron/core/ssm/mamba_mixer.py [hunk 2]
```diff
@@ -207,9 +208,10 @@ def __init__(
             self.nheads = self.d_inner // self.headdim
 
         if self.config.fp8:
-            assert (2 * self.d_inner + 2 * self.ngroups * self.d_state + self.nheads) % 16 == 0, (
+            fp8_align_size = get_fp8_align_size(self.config.fp8_recipe)
```
Comment (duncanriach) [inline megatron/core/ssm/mamba_mixer.py]: What prompts this fix in this PR?

File: tests/unit_tests/models/test_mamba_model.py [hunk 3]
```diff
@@ -394,3 +395,79 @@ def test_dynamic_inference_padding_with_fp8(self):
 
         # Assert that all padding logits are zero.
         assert torch.all(padding_logits == 0.0), "Logits for padding tokens are not all zero."
+
+
+class TestMambaBlockwiseFP8:
```
Comment (duncanriach) [inline tests/unit_tests/models/test_mamba_model.py]: Why in this PR?

File: tools/checkpoint/remap_gpt_dsa_to_mamba.py
Comment (duncanriach) [inline tools/checkpoint/remap_gpt_dsa_to_mamba.py:1]: Ugh, yeah. Checkpoint compatibility is an issue

File: megatron/core/ssm/mamba_hybrid_layer_allocation.py (hunk 1 above)
Comment (janEbert) [reply megatron/core/ssm/mamba_hybrid_layer_allocation.py]: Makes sense!

File: megatron/core/ssm/mamba_mixer.py (hunk 2 above)
Comment (janEbert) [reply megatron/core/ssm/mamba_mixer.py]: Sorry, just wanted to push it for safety's sake. I'll move the DeepSeek-related changes to another branch. :)
Didn't mean anyone to look at this yet since it's still a draft.

File: tests/unit_tests/models/test_mamba_model.py (hunk 3 above)
Comment (janEbert) [reply tests/unit_tests/models/test_mamba_model.py]: See above.

**Comment by janEbert** [comment]
Thank you for the quick review, Duncan! Should I maybe already start rebasing this on top of #3377?

File: megatron/core/ssm/mamba_mixer.py (hunk 2 above)
Comment (janEbert) [reply megatron/core/ssm/mamba_mixer.py]: Removed the unclean changes; sorry about that, I see the error in my ways and should've just pushed those to a different branch.

**Comment by duncanriach** [comment]
> Thank you for the quick review, Duncan! Should I maybe already start rebasing this on top of #3377?

Maybe good to hold off a bit since there might be some more changes in the PR now. Quite a lot of feedback just came in, including from you :-)

**Comment by janEbert** [comment]
I disabled the ability to create hybrid standard Attention+DSA models for now. This would require larger changes.

File: tests/unit_tests/ssm/test_mamba_hybrid_layer_allocation.py [hunk 4]
```diff
@@ -93,6 +96,8 @@ def test_invalid_symbols_cause_failure(self):
             validate_segment_layers("M|M")  # pipe not valid in a segment
```
[the remark this answers is not in the document]
Comment (janEbert) [reply tests/unit_tests/ssm/test_mamba_hybrid_layer_allocation.py:98]: Changed wording from "MLA" to "MLA/DSA". DSA uses MLA under the hood.

File: megatron/core/ssm/mamba_block.py [hunk 5]
```diff
@@ -134,6 +135,15 @@ def __init__(
...
                     )
+                elif layer_type == LayerSymbols.DS_ATTENTION:
+                    layer = build_module(
+                        submodules.dsa_layer,
+                        config=self.config,
+                        layer_number=layer_number,
+                        pg_collection=pg_collection,
+                        is_mtp_layer=is_mtp_layer,
+                        add_layer_offset=False,
+                    )
```
[the remark this answers is not in the document]
Comment (janEbert) [reply megatron/core/ssm/mamba_block.py:148]: Confirmed this was important using pipeline-parallel inference tests. Added in the latest revision.

File: tools/checkpoint/remap_gpt_dsa_to_mamba.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+import argparse
+import os
+import re
+import shutil
+from pathlib import Path
+from typing import Dict
+
+
+def _remap_key(key: str, num_gpt_layers: int) -> str:
```
[the remark this answers is not in the document]
Comment (janEbert) [reply tools/checkpoint/remap_gpt_dsa_to_mamba.py:45]: Started to use the variable for an assertion.

File: megatron/core/models/mamba/mamba_model.py [hunk 7]
```diff
@@ -213,7 +213,7 @@ def __init__(
                 tp_group=self.pg_collection.tp,
             )
 
-        if self.position_embedding_type == 'rope':
+        if self.position_embedding_type == 'rope' and not self.config.multi_latent_attention:
```
Comment (jaredcasper) [inline megatron/core/models/mamba/mamba_model.py:222]: Instead of silently not building rope here in this case, should we assert in the config dataclass (and/or this classes constructor) that position_embedding_type is not 'rope' if multi_latent_attention is True?

Comment (janEbert) [reply megatron/core/models/mamba/mamba_model.py:222]: The MLA layer (also used by DSA under the hood) builds its own ("decoupled") RoPE inside itself, so it's not that RoPE isn't built at all with these settings. So, semantically, I think this is sound.   This is also [how the GPTModel handles it](https://github.com/NVIDIA/Megatron-LM/blob/768855703fc78b96302f788697e55b698b4591e1/megatron/core/models/gpt/gpt_model.py#L159), so adding the suggested assertion in the config dataclass would be backward-incompatible. An alternative is implementing the assertion only in `MambaModel` like you also suggested, where it wouldn't cause backward-incompatibilities.

However, your suggestion here raises another very important point: when `position_embedding_type != 'rope' and multi_latent_attention`, this should definitely raise an error for both models! I'd submit this check as a separate PR, I guess, since it's unrelated to this implementation. WDYT?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 8]
```diff
@@ -330,16 +334,18 @@ def __init__(self, model_config: TransformerConfig, inference_config: InferenceC
             # For hybrid models, the layer map converts the global layer index to the
             # corresponding attention layer index or Mamba layer index depending on the
             # layer type.
-            mamba_layer_map, gdn_layer_map, attention_layer_map, _, _ = (
-                get_layer_maps_from_layer_type_list(mamba_inference_state_config.layer_type_list)
+            attention_layer_map, dsa_layer_map, gdn_layer_map, mamba_layer_map = (
+                operator.itemgetter(
+                    Symbols.ATTENTION, Symbols.DS_ATTENTION, Symbols.GDN, Symbols.MAMBA
+                )(get_layer_maps_from_layer_type_list(mamba_inference_state_config.layer_type_list))
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py:342]: Why do we need the operator.itemgetter wrapper now? Also similar to the GDN check, should we also be erroring if there are DSA layers?

File: tests/unit_tests/ssm/test_mamba_block.py [hunk 9]
```diff
@@ -138,3 +173,34 @@ def test_gdn_gpu_forward(self):
...
+        # layer_pattern = Symbols.MAMBA + Symbols.ATTENTION + Symbols.DS_ATTENTION + Symbols.MAMBA
+        # block = self.get_dsa_mamba_block(layer_pattern)
+        # layers = block.layers
+        # assert isinstance(layers[0], MambaLayer)
+        # assert isinstance(layers[1], TransformerLayer)
+        # assert isinstance(layers[1].self_attention, SelfAttention)
+        # assert isinstance(layers[2], TransformerLayer)
+        # assert isinstance(layers[2].self_attention, MLASelfAttention)
+        # assert isinstance(layers[2].self_attention.core_attention, DSAttention)
+        # assert isinstance(layers[3], MambaLayer)
```
Comment (santhnm2) [inline tests/unit_tests/ssm/test_mamba_block.py]: Do we have explicit plans to support this?

File: megatron/core/inference/contexts/dynamic_context.py (hunk 8 above)
Comment (janEbert) [reply megatron/core/inference/contexts/dynamic_context.py:342]: [I suggested using named structs instead of tuples back in Duncan's fVPP PR.](https://github.com/NVIDIA/Megatron-LM/pull/3377#discussion_r2851920950) This is just supposed to make the code easier to use and future-proof (because the order of the return values doesn't matter anymore, you neither have to remember the order and changes to the return value don't break existing destructuring).

As for erroring if here are DSA layers: as far as I know and have tested, DSA is supported for inference (and these changes enable it working with Mamba inference), so I would keep allowing them.

File: tests/unit_tests/ssm/test_mamba_block.py (hunk 9 above)
Comment (janEbert) [reply tests/unit_tests/ssm/test_mamba_block.py]: I don't think there's a clear decision yet. Should I remove the potentially unnecessary code for now?

**Comment by janEbert** [comment]
> Several of the new functional tests have empty values, is this expected? If the intent is to check that no inference was executed can we use unit tests instead?

[(From this comment)](https://github.com/NVIDIA/Megatron-LM/pull/3553#pullrequestreview-4072865663)

I've removed the functional tests for now and will put them in a separate PR to keep this one going.

File: tests/unit_tests/ssm/test_mamba_block.py (hunk 9 above)
Comment (santhnm2) [reply tests/unit_tests/ssm/test_mamba_block.py]: Yea I think it would be good to remove the comments now if there's a chance they would remain indefinitely.

Comment (janEbert) [reply tests/unit_tests/ssm/test_mamba_block.py]: Removed them.

File: megatron/core/models/mamba/mamba_model.py (hunk 7 above)
Comment (jaredcasper) [reply megatron/core/models/mamba/mamba_model.py:222]: Apologies for the delay getting back to you.

I think maybe just add in a comment here that the RoPE is included in the MLA layer, so not built here.

Comment (janEbert) [reply megatron/core/models/mamba/mamba_model.py:222]: Added in a8fd8a09ba84e8e37549426e25c3447d262c021d. Thank you!

**Comment by ko3n1g** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3553/commits/bb321835c40b787db645a3afa1ba71989fab47f5
